security.go

```diff
@@ -1,47 +1,91 @@
-	}
-
-	hash := hmac.New(sha256.New, []byte(signingSecret))
-	hash.Write([]byte(fmt.Sprintf("v0:%s:", header["X-Slack-Request-Timestamp"][0])))
-	return SecretsVerifier{
-		slackSig:  header["X-Slack-Signature"][0],
-		timeStamp: header["X-Slack-Request-Timestamp"][0],
-		hmac:      hash,
-	}, nil
-}
-
-func (v *SecretsVerifier) Write(body []byte) (n int, err error) {
-	return v.hmac.Write(body)
-}
-
-// Ensure compares the signature sent from Slack with the actual computed hash to judge validity
-func (v SecretsVerifier) Ensure() error {
-	computed := "v0=" + string(hex.EncodeToString(v.hmac.Sum(nil)))
-	if computed == v.slackSig {
-		return nil
-	}
-
-	return fmt.Errorf("Expected signing signature: %s, but computed: %s", v.slackSig, computed)
-}
+package slack
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// Signature headers
+const (
+	hSignature = "X-Slack-Signature"
+	hTimestamp = "X-Slack-Request-Timestamp"
+)
+
+// SecretsVerifier contains the information needed to verify that the request comes from Slack
+type SecretsVerifier struct {
+	signature string
+	hmac      hash.Hash
+}
+
+func unsafeSignatureVerifier(header http.Header, secret string) (_ SecretsVerifier, err error) {
+	signature := header.Get(hSignature)
+	stimestamp := header.Get(hTimestamp)
+
+	if signature == "" || stimestamp == "" {
+		return SecretsVerifier{}, errors.New("missing headers")
+	}
+
+	hash := hmac.New(sha256.New, []byte(secret))
+	hash.Write([]byte(fmt.Sprintf("v0:%s:", stimestamp)))
+
+	return SecretsVerifier{
+		signature: signature,
+		hmac:      hash,
+	}, nil
+}
+
+// NewSecretsVerifier returns a SecretsVerifier object in exchange for an http.Header object and signing secret
+func NewSecretsVerifier(header http.Header, secret string) (sv SecretsVerifier, err error) {
+	var (
+		timestamp int64
+	)
+
+	stimestamp := header.Get(hTimestamp)
+
+	if sv, err = unsafeSignatureVerifier(header, secret); err != nil {
+		return sv, err
+	}
+
+	if timestamp, err = strconv.ParseInt(stimestamp, 10, 64); err != nil {
+		return SecretsVerifier{}, err
+	}
+
+	diff := absDuration(time.Now().Sub(time.Unix(timestamp, 0)))
+	if diff > 5*time.Minute {
+		return sv, fmt.Errorf("timestamp is too old")
+	}
+
+	return sv, err
+}
+
+func (v *SecretsVerifier) Write(body []byte) (n int, err error) {
+	return v.hmac.Write(body)
+}
+
+// Ensure compares the signature sent from Slack with the actual computed hash to judge validity
+func (v SecretsVerifier) Ensure() error {
+	computed := "v0=" + string(hex.EncodeToString(v.hmac.Sum(nil)))
+
+	// use hmac.Equal prevent leaking timing information.
+	if hmac.Equal([]byte(computed), []byte(v.signature)) {
+		return nil
+	}
+
+	return fmt.Errorf("Expected signing signature: %s, but computed: %s", v.signature, computed)
+}
+
+func abs64(n int64) int64 {
+	y := n >> 63
+	return (n ^ y) - y
```

Comment:
Neat!

Reply:
@adamrothman this is too clever for me to understand, why we need this `absDuration` and `abs64`?

Reading Slack docs, they use

```
if absolute_value(time.time() - timestamp) > 60 * 5:
  ...
```

Where I have no clue what they mean by `absolute_value` function call since I googled it in Python land but didn't get any results back, probably there is some computer science term that I am missing.

So I will appreciate if you could explain me these two functions in this context 🙏

Reply:
The absolute value of a number, formally denoted as `|x|`, is defined as its magnitude or distance from zero. That is to say:
* `|x| = x when x > 0`
* `|x| = -x when x < 0`

In this context, the absolute value of the difference between the local system time and the request timestamp is used to allow for clock skew in either direction.

This particular implementation uses [two's complement](https://en.wikipedia.org/wiki/Two%27s_complement) arithmetic. It comes from [chapter 2 of _Hacker's Delight_ by Henry S. Warren](https://books.google.com.au/books?id=VicPJYM0I5QC&lpg=PA18&ots=2o-SROAuXq&dq=hackers%20delight%20absolute&pg=PA18#v=onepage&q=hackers%20delight%20absolute&f=false), by way of Ryan Armstrong's blog: [Optimized abs() for int64 in Go](http://cavaliercoder.com/blog/optimized-abs-for-int64-in-go.html).

The `absDuration` function just handles the type conversions to make everything a little more ergonomic.

Reply:
> The absolute value of a number, formally denoted as |x|, 

Ahhhhh that absolute value 🤦‍♂ gosh I am stupid, I thought was something else 😄 

Thanks a lot for the explanation, appreciated!